File: api/_lib/scheduler.ts
export interface Exam {
  date: string
  time: string
}

export interface Student {
  name: string
  exams: Exam[]
}

export interface LectureWindow {
  date: string
  start_time: string
  end_time: string
  available_students: number
}

export interface ScheduleResponse {
  total_students: number
  best_windows: LectureWindow[]
}

const WORK_START_MINUTES = 9 * 60
const WORK_END_MINUTES = 21 * 60
const LECTURE_DURATION_MINUTES = 3 * 60
const EXAM_DURATION_MINUTES = 90
const WINDOW_STEP_MINUTES = 30
const TOP_N = 10

interface TimeInterval {
  startMs: number
  endMs: number
}

function parseDate(value: string): Date {
  const [year, month, day] = value.split('-').map(Number)
  return new Date(Date.UTC(year, month - 1, day))
}

function parseTime(value: string): { hours: number; minutes: number } {
  for (const fmt of [/^(\d{1,2}):(\d{2})$/, /^(\d{1,2}):(\d{2}):(\d{2})$/]) {
    const match = value.match(fmt)
    if (match) {
      return { hours: Number(match[1]), minutes: Number(match[2]) }
    }
  }
  throw new Error(`Invalid time format: ${value}`)
}

function combineUtc(dateValue: Date, hours: number, minutes: number): number {
  return Date.UTC(
    dateValue.getUTCFullYear(),
    dateValue.getUTCMonth(),
    dateValue.getUTCDate(),
    hours,
    minutes,
  )
}

function addDays(dateValue: Date, days: number): Date {
  const next = new Date(dateValue)
  next.setUTCDate(next.getUTCDate() + days)
  return next
}

function formatDate(dateValue: Date): string {
  const year = dateValue.getUTCFullYear()
  const month = String(dateValue.getUTCMonth() + 1).padStart(2, '0')
  const day = String(dateValue.getUTCDate()).padStart(2, '0')
  return `${year}-${month}-${day}`
}

function formatTime(hours: number, minutes: number): string {
  return `${String(hours).padStart(2, '0')}:${String(minutes).padStart(2, '0')}`
}

function overlaps(a: TimeInterval, b: TimeInterval): boolean {
  return a.startMs < b.endMs && a.endMs > b.startMs
}

function* iterDates(start: Date, end: Date): Generator<Date> {
  let current = start
  while (current.getTime() <= end.getTime()) {
    yield current
    current = addDays(current, 1)
  }
}

function* generateDailyWindows(day: Date): Generator<TimeInterval> {
  const latestStart = WORK_END_MINUTES - LECTURE_DURATION_MINUTES

  for (
    let startMinutes = WORK_START_MINUTES;
    startMinutes <= latestStart;
    startMinutes += WINDOW_STEP_MINUTES
  ) {
    const startHours = Math.floor(startMinutes / 60)
    const startMins = startMinutes % 60
    const endMinutes = startMinutes + LECTURE_DURATION_MINUTES
    const endHours = Math.floor(endMinutes / 60)
    const endMins = endMinutes % 60

    yield {
      startMs: combineUtc(day, startHours, startMins),
      endMs: combineUtc(day, endHours, endMins),
    }
  }
}

function generateLectureWindows(startDate: string, endDate: string): TimeInterval[] {
  const start = parseDate(startDate)
  const end = parseDate(endDate)
  const windows: TimeInterval[] = []

  for (const day of iterDates(start, end)) {
    windows.push(...generateDailyWindows(day))
  }

  return windows
}

function buildExamIntervals(student: Student): TimeInterval[] {
  return student.exams.map((exam) => {
    const examDate = parseDate(exam.date)
    const { hours, minutes } = parseTime(exam.time)
    const startMs = combineUtc(examDate, hours, minutes)
    return {
      startMs,
      endMs: startMs + EXAM_DURATION_MINUTES * 60 * 1000,
    }
  })
}

function isStudentAvailable(student: Student, window: TimeInterval): boolean {
  return !buildExamIntervals(student).some((examInterval) => overlaps(examInterval, window))
}

function countAvailableStudents(students: Student[], window: TimeInterval): number {
  return students.filter((student) => isStudentAvailable(student, window)).length
}

function rankLectureWindows(students: Student[], windows: TimeInterval[]): LectureWindow[] {
  const ranked = windows.map((window) => {
    const startDate = new Date(window.startMs)
    const endDate = new Date(window.endMs)

    return {
      date: formatDate(startDate),
      start_time: formatTime(startDate.getUTCHours(), startDate.getUTCMinutes()),
      end_time: formatTime(endDate.getUTCHours(), endDate.getUTCMinutes()),
      available_students: countAvailableStudents(students, window),
    }
  })

  ranked.sort((a, b) => {
    if (b.available_students !== a.available_students) {
      return b.available_students - a.available_students
    }
    if (a.date !== b.date) {
      return a.date.localeCompare(b.date)
    }
    return a.start_time.localeCompare(b.start_time)
  })

  return ranked.slice(0, TOP_N)
}

export function findBestLectureWindows(
  students: Student[],
  startDate: string,
  endDate: string,
): ScheduleResponse {
  const windows = generateLectureWindows(startDate, endDate)
  const bestWindows = rankLectureWindows(students, windows)

  return {
    total_students: students.length,
    best_windows: bestWindows,
  }
}

File: backend/scheduler.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import date, datetime, time, timedelta
from typing import Any, Iterable, List, Sequence

WORK_START = time(9, 0)
WORK_END = time(21, 0)
LECTURE_DURATION = timedelta(hours=3)
EXAM_DURATION = timedelta(minutes=90)
WINDOW_STEP = timedelta(minutes=30)
TOP_N = 10


@dataclass(frozen=True)
class TimeInterval:
    start: datetime
    end: datetime

    def overlaps(self, other: TimeInterval) -> bool:
        return self.start < other.end and self.end > other.start


@dataclass(frozen=True)
class LectureWindow:
    date: date
    start_time: time
    end_time: time
    available_students: int

    def to_dict(self) -> dict[str, Any]:
        return {
            "date": self.date.isoformat(),
            "start_time": self.start_time.strftime("%H:%M"),
            "end_time": self.end_time.strftime("%H:%M"),
            "available_students": self.available_students,
        }


def parse_date(value: str) -> date:
    return datetime.strptime(value, "%Y-%m-%d").date()


def parse_time(value: str) -> time:
    for fmt in ("%H:%M", "%H:%M:%S"):
        try:
            return datetime.strptime(value, fmt).time()
        except ValueError:
            continue
    raise ValueError(f"Invalid time format: {value}")


def combine(date_value: date, time_value: time) -> datetime:
    return datetime.combine(date_value, time_value)


def iter_dates(start: date, end: date) -> Iterable[date]:
    current = start
    while current <= end:
        yield current
        current += timedelta(days=1)


def generate_daily_windows(day: date) -> Iterable[TimeInterval]:
    day_start = combine(day, WORK_START)
    day_end = combine(day, WORK_END)
    latest_start = day_end - LECTURE_DURATION

    current_start = day_start
    while current_start <= latest_start:
        yield TimeInterval(start=current_start, end=current_start + LECTURE_DURATION)
        current_start += WINDOW_STEP


def generate_lecture_windows(start_date: str, end_date: str) -> List[TimeInterval]:
    start = parse_date(start_date)
    end = parse_date(end_date)
    windows: List[TimeInterval] = []

    for day in iter_dates(start, end):
        windows.extend(generate_daily_windows(day))

    return windows


def build_exam_intervals(student: Any) -> List[TimeInterval]:
    intervals: List[TimeInterval] = []

    for exam in student.exams:
        exam_date = parse_date(exam.date)
        exam_start = combine(exam_date, parse_time(exam.time))
        intervals.append(TimeInterval(start=exam_start, end=exam_start + EXAM_DURATION))

    return intervals


def is_student_available(student: Any, window: TimeInterval) -> bool:
    for exam_interval in build_exam_intervals(student):
        if exam_interval.overlaps(window):
            return False
    return True


def count_available_students(students: Sequence[Any], window: TimeInterval) -> int:
    return sum(1 for student in students if is_student_available(student, window))


def rank_lecture_windows(
    students: Sequence[Any], windows: Sequence[TimeInterval]
) -> List[LectureWindow]:
    ranked: List[LectureWindow] = []

    for window in windows:
        ranked.append(
            LectureWindow(
                date=window.start.date(),
                start_time=window.start.time(),
                end_time=window.end.time(),
                available_students=count_available_students(students, window),
            )
        )

    ranked.sort(
        key=lambda item: (-item.available_students, item.date, item.start_time)
    )
    return ranked[:TOP_N]


def find_best_lecture_windows(
    students: Sequence[Any], start_date: str, end_date: str
) -> dict[str, Any]:
    windows = generate_lecture_windows(start_date, end_date)
    best_windows = rank_lecture_windows(students, windows)

    return {
        "total_students": len(students),
        "best_windows": [window.to_dict() for window in best_windows],
    }
